Search using run's attributes

mparkhe opened the pull request:
## What changes are proposed in this pull request?
 
- Added support for including run's attributes in search string
- `start_time` and `end_time` are numeric values and handle all numeric operators for search
- run's `status` is exposed as a `INT` in `RunInfo` (treating that as numeric for now) -- might need an API change to convert those to string/enum values : cc @sueann 
- tests for search and search_utils for attributes
 
## How is this patch tested?

Unit tests 
 
## Release Notes
 
### Is this a user-facing change? 

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.
 
(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)
 
### What component(s) does this PR affect?
 
- [ ] UI
- [ ] CLI 
- [x] API 
- [ ] REST-API 
- [ ] Examples 
- [ ] Docs
- [x] Tracking
- [ ] Projects 
- [ ] Artifacts 
- [ ] Models 
- [ ] Scoring 
- [ ] Serving
- [ ] R
- [ ] Java
- [ ] Python

### How should the PR be classified in the release notes? Choose one:
 
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [x] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

Commit:
Support for search using run attributes

Commit:
Merge branch 'master' of github.com:mlflow/mlflow into attr_search

Commit:
- run's status is a numeric property (RunInfo)
- adding unit tests for search using attributes

Commit:
Merge branch 'master' of github.com:mlflow/mlflow into attr_search

Commit:
lint

Commit:
Update search_utils.py

de-lint

smurching commented on mlflow/utils/search_utils.py:
```diff
@@ -10,15 +11,23 @@ class SearchFilter(object):
     _TAG_IDENTIFIER = "tag"
     _ALTERNATE_TAG_IDENTIFIERS = set(["tags"])
+    _ATTRIBUTE_IDENTIFIER = "attribute"
+    _ALTERNATE_ATTRIBUTE_IDENTIFIERS = set(["attr", "run"])
```
micronit: should we allow plural "attributes" here since we allow plural "metrics", "params", "tags"?

smurching commented on mlflow/utils/search_utils.py:
```diff
@@ -254,6 +280,22 @@ def does_run_match_clause(cls, run, sed):
+                    raise MlflowException("Invalid comparator '{}' not one of "
+                                          "'{}".format(comparator,
+                                                       cls.VALID_STRING_ATTRIBUTE_COMPARATORS))
+                lhs = getattr(run.info, key)
```
nice this is pretty cool :)

smurching commented on tests/store/test_sqlalchemy_store.py:
```diff
@@ -646,9 +646,10 @@ def test_error_logging_to_deleted_run(self):
                 run_view_type=ViewType.ALL, max_results=SEARCH_MAX_RESULTS_DEFAULT):
         search_filter = SearchFilter(filter_string=filter_string)
+        exps = [experiment_id] if isinstance(experiment_id, int) else experiment_id
+        print(self.store.search_runs(exps, search_filter, run_view_type, max_results))
```
micronit: remove the `print`? or maybe we can compute the result before printing. the real underlying issue is that `self.store.search_runs(exps, search_filter, run_view_type, max_results)` is actually somewhat slow against a SQLite-file-backed store, so tests that use this API now runs kinda slowly.

mparkhe replied:
Good catch on that `print` thing. I was debugging -- forgot to remove. We shud add `print` check in that linter!

mparkhe replied:
slow search is a legit issue. The sql push-down should help by effectively not pulling out all runs only valid ones.

smurching reviewed (approved):
Had some tiny nits but otherwise the code LGTM! Thanks for doing the follow-up work of writing out some of the tradeoffs of this filter format vs other formats for time-based search :)

Commit:
Merge branch 'master' of github.com:mlflow/mlflow into attr_search

Commit:
Merge remote-tracking branch 'mani/attr_search' into attr_search

mparkhe commented:
> Thanks for doing the follow-up work of writing out some of the tradeoffs of this filter format vs other formats for time-based search :)

Will do!

Commit:
addressing review comments and some cleanups

mparkhe commented:
@smurching : addressed your comments. Will merge in changes based on your LGTM comment. If you would for me to make any other changes, I can make a follow-on fix.
